File: 29Items-Projects/7Algorithms/2GRAPH-ALGORITHMS/java/src/main/java/com/example/graphalgorithms/BellmanFord.java
package com.example.graphalgorithms;

import java.util.*;

public class BellmanFord {
    public static class Edge {
        private final String from;
        private final String to;
        private final int weight;

        public Edge(String from, String to, int weight) {
            this.from = from;
            this.to = to;
            this.weight = weight;
        }

        public String from() { return from; }
        public String to() { return to; }
        public int weight() { return weight; }
        public String toString() { return "(" + from + " -> " + to + ", " + weight + ")"; }
    }

    public static Map<String, Integer> shortestPaths(List<String> vertices, List<Edge> edges, String start) {
        if (!vertices.contains(start)) {
            throw new IllegalArgumentException("start vertex is not in the graph");
        }
        if (new HashSet<>(vertices).size() != vertices.size()) {
            throw new IllegalArgumentException("vertices must be unique");
        }
        Map<String, Integer> distance = new HashMap<>();
        for (String vertex : vertices) {
            distance.put(vertex, Integer.MAX_VALUE / 4);
        }
        distance.put(start, 0);

        for (int i = 0; i < vertices.size() - 1; i++) {
            boolean changed = false;
            for (Edge edge : edges) {
                if (!distance.containsKey(edge.from()) || !distance.containsKey(edge.to())) {
                    throw new IllegalArgumentException("edge references an unknown vertex");
                }
                int fromDistance = distance.get(edge.from());
                if (fromDistance < Integer.MAX_VALUE / 4 && fromDistance + edge.weight() < distance.get(edge.to())) {
                    distance.put(edge.to(), fromDistance + edge.weight());
                    changed = true;
                }
            }
            if (!changed) {
                break;
            }
        }
        for (Edge edge : edges) {
            if (!distance.containsKey(edge.from()) || !distance.containsKey(edge.to())) {
                throw new IllegalArgumentException("edge references an unknown vertex");
            }
            int fromDistance = distance.get(edge.from());
            if (fromDistance < Integer.MAX_VALUE / 4 && fromDistance + edge.weight() < distance.get(edge.to())) {
                throw new IllegalArgumentException("Graph contains a negative weight cycle");
            }
        }
        return distance;
    }

    public static void main(String[] args) {
        System.out.println("Bellman-Ford shortest paths from A");
        System.out.println(shortestPaths(
            List.of("A", "B", "C"),
            List.of(new Edge("A", "B", 4), new Edge("A", "C", 5), new Edge("B", "C", -2)),
            "A"
        ));
    }
}

File: 29Items-Projects/7Algorithms/2GRAPH-ALGORITHMS/python/graph_algorithms/bellman_ford.py
from __future__ import annotations

from math import inf


def bellman_ford(vertices: list[str], edges: list[tuple[str, str, int]], start: str) -> dict[str, float]:
    if start not in vertices:
        raise ValueError(f"start vertex {start!r} is not in the graph")
    if len(vertices) != len(set(vertices)):
        raise ValueError("vertices must be unique")
    distance = {vertex: inf for vertex in vertices}
    distance[start] = 0

    for _ in range(len(vertices) - 1):
        changed = False
        for source, target, weight in edges:
            if source not in distance or target not in distance:
                raise ValueError("edge references an unknown vertex")
            if distance[source] != inf and distance[source] + weight < distance[target]:
                distance[target] = distance[source] + weight
                changed = True
        if not changed:
            break

    for source, target, weight in edges:
        if source not in distance or target not in distance:
            raise ValueError("edge references an unknown vertex")
        if distance[source] != inf and distance[source] + weight < distance[target]:
            raise ValueError("Graph contains a negative weight cycle")
    return distance


if __name__ == "__main__":
    nodes = ["A", "B", "C", "D"]
    weighted_edges = [("A", "B", 4), ("A", "C", 5), ("B", "C", -2), ("C", "D", 3)]
    print("Bellman-Ford shortest paths from A")
    print(bellman_ford(nodes, weighted_edges, "A"))
